enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/id/ReplicatedIdGenerator.java

```diff
@@ -34,27 +39,38 @@
     private final IdType idType;
     private final Log log;
     private final ReplicatedIdRangeAcquirer acquirer;
+    private final BooleanSupplier freeIdCondition;
     private volatile long highId;
-    private volatile long defragCount;
     private volatile IdRangeIterator idQueue = EMPTY_ID_RANGE_ITERATOR;
 
-    ReplicatedIdGenerator( IdType idType, long highId, ReplicatedIdRangeAcquirer acquirer, LogProvider
-            logProvider )
+    private IdContainer idContainer;
+
+    ReplicatedIdGenerator( FileSystemAbstraction fs, File file, IdType idType, long highId,
+            ReplicatedIdRangeAcquirer acquirer, LogProvider logProvider, int grabSize, boolean aggressiveReuse,
+            BooleanSupplier freeIdCondition )
     {
         this.idType = idType;
         this.highId = highId;
         this.acquirer = acquirer;
         this.log = logProvider.getLog( getClass() );
+        this.freeIdCondition = freeIdCondition;
+        idContainer = new IdContainer( fs, file, grabSize, aggressiveReuse );
+        idContainer.init();
     }
 
     @Override
     public void close()
     {
+        idContainer.close( highId );
     }
 
     @Override
     public void freeId( long id )
     {
+        if ( freeIdCondition.getAsBoolean() )
+        {
+            idContainer.freeId( id );
+        }
```

Comment:
not sure, I fully understand your question. But in case I am: for period when new master catching up with transaction which already in the raft log but not applied yet